### DISCUSSION TITLE
Is elektra plattform independent?

### DISCUSSION THREAD

**Discussion by buhtz** [ElektraInitiative/libelektra#4081] [General]
Today I read about the new Elektra release.
As a developer of a Python3 application I wonder if I could support the project with using Elektra instead of simple json-files somewhere in `$XDG_CONFIG_HOME`.

But what is about platforms that are not linux/unix? e.g. Windows, Haiku, ReactOS, etc.
Is there a way for Elektra?

Or is there a backend/plugin for Elektra that could be used on plattforms where Elektra is not available (by default). I then could "bind" that plugin to my non-linux version of the application. You could call it an electra dummy.

Python like pseudocode:

```
try:
    # the real elektra (on linux)
    import elektra
except ImportError:
   # a dummy offering the usual elektra api
   # e.g. on windows it would read/write config data in $APP_DATA
   import ElektraDummy as elektra

```

**Comment by markus2330** [comment]
Dear Buhtz,

> Today I read about the new Elektra release.

:rocket: Can you tell us where you read about it?

> if I could support the project with using Elektra

Absolutely :sparkling_heart: Also any issue you report or fix is of great help!

> But what is about platforms that are not linux/unix? e.g. Windows, Haiku, ReactOS, etc.
> Is there a way for Elektra?

Yes, definitely. Operating-system-specific stuff is exclusively implemented in plugins, the core is pure portable C (there will also be a Rust reimplementation soon). These operating-system-specific plugins use POSIX, so all POSIX systems should work without any hassle. Non-POSIX systems work by reimplementing the plugins. For, example, for Windows we have an alternative resolver https://www.libelektra.org/plugins/wresolver which is probably portable enough to also work on other non-POSIX systems.

Redshift with Elektra https://github.com/jonls/redshift/pull/837 is reported to work nicely on Windows. Only some non-essential features like some validation plugins are not working.

> try:   import elektra

I did this already often and it works quite well but it has the obvious disadvantage that you have two different code-paths to maintain. For migration until Elektra is more commonly available (e.g. 1.0 release) it might be a good compromise. We do not deliver such "DummyElektra", though, so you would need to implement it yourself. In your case, you could probably simply use the old JSON code you already have during the migration period. Our goal is that such fallback code will not be necessary.

**Comment by buhtz** [comment]
Hello Markus,

thanks for the quick reply.

I checked the python examples in the language bindings section of your 
documentation. I am not sure if I get it how it works. But I am not able 
to test it myself.
There is no package in Debian or PyPi or somewhere else.

How can I check it out in Python?

Kind
Christian

Am 06.10.2021 18:02 schrieb Markus Raab:
> Dear Buhtz,
> 
>> Today I read about the new Elektra release.
> 
> :rocket: Can you tell us where you read about it?
> 
>> if I could support the project with using Elektra
> 
> Absolutely :sparkling_heart: Also any issue you report or fix is of 
> great help!
> 
>> But what is about platforms that are not linux/unix? e.g. Windows, 
>> Haiku, ReactOS, etc.
>> Is there a way for Elektra?
> 
> Yes, definitely. Operating-system-specific stuff is exclusively
> implemented in plugins, the core is pure portable C (there will also
> be a Rust reimplementation soon). These operating-system-specific
> plugins use POSIX, so all POSIX systems should work without any
> hassle. Non-POSIX systems work by reimplementing the plugins. For,
> example, for Windows we have an alternative resolver
> https://www.libelektra.org/plugins/wresolver which is probably
> portable enough to also work on other non-POSIX systems.
> 
> Redshift with Elektra https://github.com/jonls/redshift/pull/837 is
> reported to work nicely on Windows. Only some non-essential features
> like some validation plugins are not working.
> 
>> try:   import elektra
> 
> I did this already often and it works quite well but it has the
> obvious disadvantage that you have two different code-paths to
> maintain. For migration until Elektra is more commonly available (e.g.
> 1.0 release) it might be a good compromise. We do not deliver such
> "DummyElektra", though, so you would need to implement it yourself. In
> your case, you could probably simply use the old JSON code you already
> have during the migration period. Our goal is that such fallback code
> will not be necessary.

**Comment by markus2330** [comment]
>There is no package in Debian or PyPi or somewhere else.

We provide Debian packages, see: https://www.libelektra.org/installation/installation

> How can I check it out in Python?

You found a bug! Somehow the python package dropped from being built, I created #4082

It should work by simply running `import kdb` after you installed the elektra python package.
